libraries/AP_RangeFinder/AP_RangeFinder_SRF02I2C.cpp

```diff
@@ -0,0 +1,162 @@
+            , _dev(std::move(dev)){}
+
+/**
+ * Detect
+ *
+ * @return nullptr is Sensor not found or Initialization failure. <BR>
+ * Other is Instance.
+ */
+AP_RangeFinder_Backend *AP_RangeFinder_SRF02I2C::detect(RangeFinder::RangeFinder_State &_state, AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    AP_RangeFinder_SRF02I2C *sensor = new AP_RangeFinder_SRF02I2C(_state, std::move(dev));
+
+    // Device initialization
+    if (!sensor || !sensor->_init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+/**
+ * update the state of the sensor
+ *
+ */
+void AP_RangeFinder_SRF02I2C::update(void)
+{
+    if (_sem->take_nonblocking()) {
+        if (_new_distance) {
+            // Measurement valid
+            state.distance_cm = _distance;
+            _new_distance = false;  // Set invalid measurement
+            update_status();
+        } else {
+            // Invalid measurement
+            set_status(RangeFinder::RangeFinder_NoData);
+        }
+        _sem->give();
+    }
+}
+
+/**
+ * Initialization
+ *
+ * @retval true Successful completion
+ * @retval false Abnormal termination
+ */
+bool AP_RangeFinder_SRF02I2C::_init()
+{
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        return false;
+    }
+
+    // Issuing measurement start command
+    bool ret = _startMeasurement();
+
+    _dev->get_semaphore()->give();
+
+    if (ret) {
+        // call timer() at 14Hz
+        _dev->register_periodic_callback(70000,
+                                         FUNCTOR_BIND_MEMBER(&AP_RangeFinder_SRF02I2C::_timer, void));
+    }
+
+    return ret;
+}
+
+/**
+ * Instruct start of measurement
+ *
+ * @retval true Successful measurement start instruction
+ * @retval false Measurement start instruction failure (with retry)
+ */
+bool AP_RangeFinder_SRF02I2C::_startMeasurement(void)
+{
+    // Set Real range mode and Set unit cm
+    static const uint8_t cmd[2] = {SRF02_I2C_REG_W_COMMAND, SRF02_I2C_CMD_REAL_RANGE_MODE_RESULT_IN_CENTIMETERS};
+
+    _dev->set_retries(10);
+    return _dev->transfer(cmd, sizeof(cmd), nullptr, 0);
+}
+
+/**
+ * Get measured distance
+ *
+ * @param [out] outMeasuredDistance measured distance
+ * @retval true Measurement normal
+ * @retval false Measurement abnormality
+ */
+bool AP_RangeFinder_SRF02I2C::_getMeasuredDistance(uint16_t &outMeasuredDistance)
+{
+    uint8_t distance[2];
+
+    uint8_t startReg = SRF02_I2C_REG_R_RANGE_HIGH_BYTE;
+    bool ret = _dev->transfer(&startReg, 1, distance, sizeof(distance));
+    if (ret) {
+        outMeasuredDistance = distance[0] << 8 | distance[1];
+        if (SRF02_I2C_MIN_DISTANCE > outMeasuredDistance) {
+            // The measured value is the lower limit
+            outMeasuredDistance = 0;
+        } else if (SRF02_I2C_MAX_DISTANCE < outMeasuredDistance) {
+            // The measured value is the higher limit
+            outMeasuredDistance = SRF02_I2C_MAX_DISTANCE;
```

Comment:
This is just wrong. If the returned distance is outside the limits, then we don't have a reading, it's not zero or max. It should simply set ret to false.

Reply:
I will change the return value.